### ISSUE TITLE
Container specific notifications

### ISSUE THREAD

**Issue by eiri020** [fmartinou/whats-up-docker#34]
As started this discussion at https://community.home-assistant.io/t/whats-up-docker-how-to-keep-your-containers-up-to-date/168315/16, I would like to monitor containers within Home Assistant. 

When a monitoring solution like WUD only sends notifications for all containers to a single MQTT topic or WebHook endpoint (for example the new trigger based sensors, https://www.home-assistant.io/blog/2021/04/07/release-20214/#trigger-based-template-sensors), it is harder to implement. With MQTT for example, only the last notification will remain as the sensor value, and so you need to use automation triggers to set values to sensors, variables, inputs or other persistent properties.

My use case is to have binary sensors based on the container name, which shows when a container needs an update. I can then manually call a simple script from HA to pull the image and restart the container (a nice but less important feature) .

It would therefore nice to have some feature, to specify which MQTT topic or WebHook to call based on the name of the container. A more enhanced version might be a templated specification of the topic message or WebHook payload, where you can specify which properties you include in the body. But again, only the first feature request would help already a lot.

Another solution might be, that you always send the notification for all containers with the update state, then you could use the MQTT discovery mechanism to create the sensors automatically, even when you add some new containers.

Not sure how exactly to implement that with with webhooks, although the new trigger based template sensors might help a bit.

Hope this helps. Looking forward to see progress or give additional input.

**Comment by fmartinou** [comment]
Hi @eiri020 ,

I have a first implementation proposal to get wud integrated into hass.
I'm focusing on MQTT integration, which is IMHO the best solution.

The related documentation [is available here](https://github.com/fmartinou/whats-up-docker/blob/feature/%2334/docs/triggers/mqtt/README.md#home-assistant-integration) (not yet publicly published).

To sum up, all monitored images can now be automatically discovered by hass ([see the official hass mqtt discovery doc](https://www.home-assistant.io/docs/mqtt/discovery/)).

Each image is a `boolean_sensor` entity indicating if a new version of the image is available.

It's already available in a release candidate version; could you give a try to the `fmartinou/whats-up-docker:develop` version ?

Any feedback will be appreciated.
Thanks!

**Comment by eiri020** [comment]
Hi fmartinou,

Very nice work! I was hoping you would choose this solution, because for Home Assistant it seems to be the best integration. 

I did a qucik test and it basically does the job and sensors are added automatically. I have a few findings, but maybe my setup is a bit too complex for this initial improvements:

Issue 1) Multiple docker hosts

I have two servers with docker running, both I want to monitor.
I prefer running the containers with a docker compose-file, which looked like this:

`
version: '3'
services:
  whats-up-docker:
    container_name: wud
    image: fmartinou/whats-up-docker:develop 
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    ports:
      - "3000:3000"
    environment:
      - TZ=Europe/Amsterdam
      - WUD_TRIGGER_MQTT_MOSQUITTO_URL=mqtt://192.168.64.10:1883
      - WUD_TRIGGER_MQTT_MOSQUITTO_HASS_ENABLED=true
      - WUD_TRIGGER_MQTT_MOSQUITTO_TOPIC=magelang/image
      - WUD_TRIGGER_MQTT_MOSQUITTO_USER=******
      - WUD_TRIGGER_MQTT_MOSQUITTO_PASSWORD=*****

    restart: always
`
On the other server I had a similar compose file, with only the base topic different. 

Now the binary sensors in HA look like:
![image](https://user-images.githubusercontent.com/10806142/116305562-3d4f1480-a7a4-11eb-8d7d-21ce65cfd0d2.png)

As you can see, I cannot distinguish the server where the image is out of date.
Preferable it would name something like binary_sensor.wud_<servername>_<container>.
Maybe  container name or image name arguable, but i personally prefer the container name, because this is where I need to run the update (in my case the docker-compose pull)

2) More info about the update
It would be nice if you could add a few more details from the message that you post to mqtt add to the binary_sensor as attribute and also add the docker host name as attribute

3) Not all images are listed
If I recall well, you mentioned this already somewhere in your documentation, but not all images are listed. I have the standard phpmyadmin container running, and this is not listed

4) After change of mqtt base topic , not all containers were listed anymore
At first I used the default topic, but when i configured the second docker host, I decided to use different topics for each host. After I restarted only your image was listed in the new topic (I noticed that you updated the image on docker hub during that time), so I think only new images were listed and the others were not updated in mqtt anymore,

Hope this feedback helps and again. great improvements so far. Would love to see this going into final release

Regards,
Erwin

**Comment by eiri020** [comment]
No sure about that 3), I see it is listed in the sensors. Strange I do not see it in the mqtt explorer:

![image](https://user-images.githubusercontent.com/10806142/116307987-3e357580-a7a7-11eb-81f6-dbc138c2f7a7.png)

**Comment by fmartinou** [comment]
> => You can also see that images that are used on both hosts (portainer, whats-up-docker), are only listed once

I think I know why:
Every entity on HASS must have a unique id.
Currently, I define the id using the following pattern: `wud_${registry_name}_${image_name}`.
This is why you get only 1 sensor (both overlap each other).

I can change the way of how the id is built.
Maybe the id could be the topic name itself? (just replacing `/` with `_` to comply with the allowed characters).
which will give you entities named as:
- magelang_image_portainer_portainer-ce
- medan_image_portainer_portainer-ce

What do you think?

**Comment by fmartinou** [comment]
> More info about the update
It would be nice if you could add a few more details from the message that you post to mqtt add to the binary_sensor as attribute and also add the docker host name as attribute

I already publish all data I have into the message ([example](https://github.com/fmartinou/whats-up-docker/tree/feature/%2334/docs/triggers/mqtt#example-of-sent-message))

And all of them are parsed by hass into sensor attributes; which one are missing?
![image](https://user-images.githubusercontent.com/16061231/116368939-2262bb80-a809-11eb-89a9-2763bf05e6c4.png)

>  also add the docker host name as attribute
You should already be able to get it using the `watcher` attribute on HASS.

If you want to distinguish your 2 Docker hosts, you could configure wud to get 2 different watcher names ([see related doc here](https://fmartinou.github.io/whats-up-docker/#/watchers/)).

For example,
On magelang you run wud with env var
```
WUD_WATCHER_MAGELANG_SOCKET=/var/run/docker.sock
```

And on medan, you run wud with env var
```
WUD_WATCHER_MEDAN_SOCKET=/var/run/docker.sock
```

**Comment by fmartinou** [comment]
> After change of mqtt base topic , not all containers were listed anymore
At first I used the default topic, but when i configured the second docker host, I decided to use different topics for each host. After I restarted only your image was listed in the new topic (I noticed that you updated the image on docker hub during that time), so I think only new images were listed and the others were not updated in mqtt anymore,

Every time WUD scans the running containers, it sends a discovery message so that hass can create the appropriate sensors.
I don't know what is could have happened...

Maybe the fact that you have 2 hosts overlapping each other is the cause of this side effect?

**Comment by fmartinou** [comment]
>  images that are used on both hosts (portainer, whats-up-docker), are only listed once

For your information, for both wud and portainer, you could run them only once (on the host of your choice) and let them access the other hosts using the [Docker tcp api](https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-socket-option).

- WUD supports it ([see wud doc here](https://fmartinou.github.io/whats-up-docker/#/watchers/?id=watch-remote-docker-host-on-2376-port-every-hour))
- Portainer also supports it [see portainer doc here](https://documentation.portainer.io/v2.0/endpoints/docker/)

It's just a suggestion ;)

**Comment by eiri020** [comment]
> @eiri020
> 
> > => You can also see that images that are used on both hosts (portainer, whats-up-docker), are only listed once
> 
> I think I know why:
> Every entity on HASS must have a unique id.
> Currently, I define the id using the following pattern: `wud_${registry_name}_${image_name}`.
> This is why you get only 1 sensor (both overlap each other).
> 
> I can change the way of how the id is built.
> Maybe the id could be the topic name itself? (just replacing `/` with `_` to comply with the allowed characters).
> which will give you entities named as:
> 
> * magelang_image_portainer_portainer-ce
> * medan_image_portainer_portainer-ce
> 
> What do you think?

This would certainly improve a lot. What do you think about using the container name instead of the image name? You could have two containers running with different image version. When the container name is visible, you know exactly which container to update. Also this is what the source is of your update scan.

**Comment by eiri020** [comment]
> > More info about the update
> > It would be nice if you could add a few more details from the message that you post to mqtt add to the binary_sensor as attribute and also add the docker host name as attribute
> 
> I already publish all data I have into the message ([example](https://github.com/fmartinou/whats-up-docker/tree/feature/%2334/docs/triggers/mqtt#example-of-sent-message))

Maybe I got a previous version. I updated now, but I guess I have to wait until next update cycle

![image](https://user-images.githubusercontent.com/10806142/116382841-9b1c4480-a816-11eb-9dd2-c839536ab889.png)

**Comment by eiri020** [comment]
> > After change of mqtt base topic , not all containers were listed anymore
> > At first I used the default topic, but when i configured the second docker host, I decided to use different topics for each host. After I restarted only your image was listed in the new topic (I noticed that you updated the image on docker hub during that time), so I think only new images were listed and the others were not updated in mqtt anymore,
> 
> Every time WUD scans the running containers, it sends a discovery message so that hass can create the appropriate sensors.
> I don't know what is could have happened...
> 
> Maybe the fact that you have 2 hosts overlapping each other is the cause of this side effect?

I suppose, I did something wrong here. I will report it when it happens again

**Comment by fmartinou** [comment]
> > > More info about the update
> > > It would be nice if you could add a few more details from the message that you post to mqtt add to the binary_sensor as attribute and also add the docker host name as attribute
> > 
> > 
> > I already publish all data I have into the message ([example](https://github.com/fmartinou/whats-up-docker/tree/feature/%2334/docs/triggers/mqtt#example-of-sent-message))
> 
> Maybe I got a previous version. I updated now, but I guess I have to wait until next update cycle
> 
> ![image](https://user-images.githubusercontent.com/10806142/116382841-9b1c4480-a816-11eb-9dd2-c839536ab889.png)

My bad.
I probably caused a regression because all attributes were sent and parsed by hass and it's indded not the case anymore.
=> I'll fix it.

**Comment by fmartinou** [comment]
> > @eiri020
> > > => You can also see that images that are used on both hosts (portainer, whats-up-docker), are only listed once
> > 
> > 
> > I think I know why:
> > Every entity on HASS must have a unique id.
> > Currently, I define the id using the following pattern: `wud_${registry_name}_${image_name}`.
> > This is why you get only 1 sensor (both overlap each other).
> > I can change the way of how the id is built.
> > Maybe the id could be the topic name itself? (just replacing `/` with `_` to comply with the allowed characters).
> > which will give you entities named as:
> > 
> > * magelang_image_portainer_portainer-ce
> > * medan_image_portainer_portainer-ce
> > 
> > What do you think?
> 
> This would certainly improve a lot. What do you think about using the container name instead of the image name? You could have two containers running with different image version. When the container name is visible, you know exactly which container to update. Also this is what the source is of your update scan.

Using container_names would be a big change regarding how things are working now.
WUD scans containers but map them to image versions before looking for updates.
The goal was to avoid to get notified multiple times when you run multiple containers from the same image version (example multiple mariadb containers sharing the same image version).

Your need is the opposite; you want to precisely know which container has to be updated.

=> I will think about it. You probably right;  working on a per-container basis is maybe the simplest thing to do.

**Comment by fmartinou** [comment]
> > > > More info about the update
> > > > It would be nice if you could add a few more details from the message that you post to mqtt add to the binary_sensor as attribute and also add the docker host name as attribute
> > > 
> > > 
> > > I already publish all data I have into the message ([example](https://github.com/fmartinou/whats-up-docker/tree/feature/%2334/docs/triggers/mqtt#example-of-sent-message))
> > 
> > 
> > Maybe I got a previous version. I updated now, but I guess I have to wait until next update cycle
> > ![image](https://user-images.githubusercontent.com/10806142/116382841-9b1c4480-a816-11eb-9dd2-c839536ab889.png)
> 
> My bad.
> I probably caused a regression because all attributes were sent and parsed by hass and it's indeed not the case anymore.
> => I'll fix it.

**FIXED**

**Comment by eiri020** [comment]
Thanks again. Then that might be a new feature request. Your current solution with image names would already be a major help. If you would add the naming based on topic like you suggested, then this will even help more to distinguish on host for the updates.

**Comment by fmartinou** [comment]
> If you would add the naming based on topic like you suggested, then this will even help more to distinguish on host for the updates.

**It should be working now**
Could you give it a try? Thanx!

**Comment by eiri020** [comment]
Super. Looks very nice. This is a version that beats the competition, when using Home Assistant.

Maybe a minor issue: Why do you compare the local image tag with the latest tag. For example, now that I installed your develop version, it compares it with the latest version on the registry, and reports an update. 
But again, this version will do the job for me.

Great update

**Comment by fmartinou** [comment]
Nice!
So let's release a `3.5.0`!

>  Why do you compare the local image tag with the latest tag. For example, now that I installed your develop version, it compares it with the latest version on the registry, and reports an update.

Actually, the primary purpose of WUD is managing **semver version**.
WUD is not very relevant when falling back to **non** semver versions.
I don't have the exact algorithm in mind but when the tag is not semver, I guess it falls back to an alphabetical comparison so **l**atest is higher than **d**evelop.

By the way, on the hass community, [@finity asked for a feature allowing to detect when `latest` points to a newer version](https://community.home-assistant.io/t/whats-up-docker-how-to-keep-your-containers-up-to-date/168315/29?u=fmartinou).

**Comment by fmartinou** [comment]
Release `3.5.0` is out!
Enjoy :)

**Comment by eiri020** [comment]
Congrats, Looks good. I will enjoy it. Thanks for implementing this.

**Comment by fmartinou** [comment]
@eiri020 ,

Please notice that now WUD is also supporting non semver tags (e.g `latest`) 🎉 .
So it should fix your issue of the bad comparison between `develop` and `latest`.

It's still in beta stage but you can already give it a try if you're interested (`fmartinou/whats-up-docker:develop` version).

To explain a little bit:
- WUD now compares the digest of the running image with the digest of the image manifest from the Registry
- If the digests are not equal, WUD reports that the image can be updated

Thanks for your feedback!

*N.B. The version is non backward compatible => Please remove /store/wud.json before upgrading*

**Comment by fmartinou** [comment]
It's now available in version `4.0.0`.

**Comment by eiri020** [comment]
Cool! Thank you. Again a nice feature. Will pull the image soon.

Thanks

**Comment by eiri020** [comment]
If you don't mind, may I come back on that reporting the container name in stead of only the image. 

I really enjoy your updates and am figuring out how to use it other then only reporting the state.
My goal will be to have status view and then perform actions on them (pull, restart, etc). Of course this out of scope for the functionality of WUD, but the only thing  I would need is a name of the container to perform the actions with some scripting.

I may not be a bad thing to have images as base, but the original list of images to scan is also based on the containers. And you can have two same images/different tags running.  Maybe your new tag feature will help here a bit.

In my workflow, I get following information of WUD with your new updates:
![image](https://user-images.githubusercontent.com/10806142/117187665-71e25200-addc-11eb-95b4-340272c7a7df.png)

So then the update cycle would be pull new image, restart server. But for restarting I need to know the container to restart. Now I can solve this now by creating a script that run through all containers and comparing the image name and tag. But it would be another nice feature to have this link to help automation in HA or other connected systems.

Excuse me for my persistent request, which should probably be just a new feature request. I wrote it here because it was mentioned. 

Again, thanks for the cool updates. It help me a lot to keep track of my containers.

**Comment by fmartinou** [comment]
Your request has not been forgotten :)

I plan to deeply rewrite the app to focus all on `containers` instead of `images`.
It will be a big (non backward compatible) change... and it will take some time.

In the meantime, as a quick fix, I can just add the container name to the other existing properties (os, arch...) so it should help you implementing what you're lookoing for.

Would that suit you?

**Comment by eiri020** [comment]
Yes that would help a lot and simplify my use case. Then I have the link and can connect container actions.

**Comment by fmartinou** [comment]
You can upgrade to version [`4.0.1`](https://fmartinou.github.io/whats-up-docker/#/CHANGELOG?id=_401) which includes the property `containerName` as expected.

**Comment by eiri020** [comment]
Very cool. Thank you so much. That was really the last obstacle in my use case. 

Looking forward to use it in Home Assistant to automate or visualize container state.
Thanks for all these great updates

**Comment by fmartinou** [comment]
Hi @eiri020 ,

> What do you think about using the container name instead of the image name? You could have two containers running with different image version. When the container name is visible, you know exactly which container to update. Also this is what the source is of your update scan.

This is now how WUD is working (starting from version `5.0.0`).
Updates are reported **per container** (and not per image as before).
Hope it will help you :)
